src/Tools/dotnet-monitor/Egress/S3/MultiPartStream.cs

```diff
@@ -0,0 +1,145 @@
+﻿using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Amazon.S3;
+using Amazon.S3.Model;
+
+namespace Microsoft.Diagnostics.Tools.Monitor.Egress.S3;
+
+internal class MultiPartUploadStream : Stream
+{
+    private readonly byte[] _buffer;
+    private int _offset;
+    private readonly string _bucketName;
+    private readonly string _objectKey;
+    private readonly string _uploadId;
+    private readonly IAmazonS3 _client;
+    private readonly List<PartETag> _parts = new();
+    public List<PartETag> Parts => _parts.ToList();
+    public bool Disposed { get; private set; }
+    private int _position;
+    private const int MinimumSize = 5 * 1024 * 1024; // the minimum size of an upload part (except for the last part)
+
+    public MultiPartUploadStream(IAmazonS3 client, string bucketName, string objectKey, string uploadId, int bufferSize)
+    {
+        bufferSize = Math.Max(bufferSize, MinimumSize); // has to be at least the minimum
+        _buffer = ArrayPool<byte>.Shared.Rent(bufferSize);
+        _offset = 0;
+        _client = client;
+        _bucketName = bucketName;
+        _objectKey = objectKey;
+        _uploadId = uploadId;
+    }
+
+    public override async Task FlushAsync(CancellationToken cancellationToken)
+    {
+        if (Disposed)
+            throw new ObjectDisposedException("The stream was already disposed");
+        if (_offset == 0 || _offset < MinimumSize)
+            return;
+        await DoWriteAsync(cancellationToken);
+    }
+
+    public async Task FinalizeAsync(CancellationToken cancellationToken)
+    {
+        if (Disposed)
+            throw new ObjectDisposedException("The stream was already disposed");
+        if (_offset == 0)
+            return;
+        await DoWriteAsync(cancellationToken);
+    }
+
+    public override void Flush()
+    {
+        FlushAsync(CancellationToken.None).GetAwaiter().GetResult();
+    }
+
+    public override int Read(byte[] buffer, int offset, int count)
+    {
+        throw new NotSupportedException();
+    }
+
+    public override long Seek(long offset, SeekOrigin origin)
+    {
+        throw new NotSupportedException();
+    }
+
+    public override void SetLength(long value)
+    {
+        throw new NotSupportedException();
+    }
+
+    public override async Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+    {
+        if (Disposed)
+            throw new ObjectDisposedException("The stream was already disposed");
+
+        for (var idx = offset; idx < offset + count; idx++)
```

Comment:
Rather than copy byte-by-byte, use some variation of [Array.Copy](https://docs.microsoft.com/en-us/dotnet/api/system.array.copy?view=net-6.0#system-array-copy(system-array-system-int32-system-array-system-int32-system-int32))

Reply:
In general I would agree but don't see a significant reason why the one or the other is better. I prefer the used approach as it is more readable and understandable.

Reply:
For performance reasons.

Egress providers have several memory-bloating the characteristics if they do not move data off of the diagnostic communication channel expeditiously. They can cause the tool's memory to significantly increase due to not pulling data off the channel, the channel could throttle causing event data to be dropped from the target process (e.g. give incomplete traces), and can cause the target process event buffer to significantly increase memory pressure. The first and last could easily cause container eviction due to memory limits being exceeded.

Byte-by-byte is possibly the slowest way to do this. Let the framework do its best job to copy as much data as efficiently as possible.